http/src/handler.rs

```diff
@@ -70,87 +63,227 @@ impl<M: Metadata, S: Middleware<M>> server::Handler<HttpStream> for ServerHandle
-		self.handler.on_response_writable(encoder)
+	fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+		match *self {
+			Handler::Rpc(ref mut handler) => handler.poll(),
+			Handler::Middleware(ref mut middleware) => middleware.poll(),
+			Handler::Error(ref mut response) => Ok(Async::Ready(
+				response.take().expect("Response always Some initialy. Returning `Ready` so will never be polled again; qed").into()
+			)),
+		}
 	}
 }
 
-enum Handler<M: Metadata, S: Middleware<M>> {
-	Rpc(RpcHandler<M, S>),
-	Error(Response),
-	Middleware(Box<server::Handler<HttpStream> + Send>),
+enum RpcPollState<M> {
+	Ready(RpcHandlerState<M>),
+	NotReady(RpcHandlerState<M>),
 }
 
-impl<M: Metadata, S: Middleware<M>> Deref for Handler<M, S> {
-	type Target = server::Handler<HttpStream>;
-
-	fn deref(&self) -> &Self::Target {
-		match *self {
-			Handler::Rpc(ref handler) => handler,
-			Handler::Error(ref handler) => handler,
-			Handler::Middleware(ref response) => &**response,
+impl<M> RpcPollState<M> {
+	fn decompose(self) -> (RpcHandlerState<M>, bool) {
+		use self::RpcPollState::*;
+		match self {
+			Ready(handler) => (handler, true),
+			NotReady(handler) => (handler, false),
 		}
 	}
 }
+enum RpcHandlerState<M> {
+	ReadingHeaders {
+		request: server::Request,
+		cors_domains: CorsDomains,
+		continue_on_invalid_cors: bool,
+	},
+	ReadingBody {
+		body: hyper::Body,
+		request: Vec<u8>,
+		metadata: M,
+	},
+	Writing(Response),
+	Waiting(BoxFuture<Option<String>, ()>),
+	Done,
+}
+
+impl<M> fmt::Debug for RpcHandlerState<M> {
+	fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+		use self::RpcHandlerState::*;
 
-impl<M: Metadata, S: Middleware<M>> DerefMut for Handler<M, S> {
-	fn deref_mut(&mut self) -> &mut Self::Target {
 		match *self {
-			Handler::Rpc(ref mut handler) => handler,
-			Handler::Error(ref mut handler) => handler,
-			Handler::Middleware(ref mut response) => &mut **response,
+			ReadingHeaders {..} => write!(fmt, "ReadingHeaders"),
+			ReadingBody {..} => write!(fmt, "ReadingBody"),
+			Writing(ref res) => write!(fmt, "Writing({:?})", res),
+			Waiting(_) => write!(fmt, "Waiting"),
+			Done => write!(fmt, "Done"),
 		}
 	}
 }
 
 pub struct RpcHandler<M: Metadata, S: Middleware<M>> {
-	cors_domains: Option<Vec<cors::AccessControlAllowOrigin>>,
 	jsonrpc_handler: Rpc<M, S>,
-	metadata: Option<M>,
-	control: Control,
-	request: Request,
-	response: Response,
-	continue_on_invalid_cors: bool,
-	/// Asynchronous response waiting to be moved into `response` field.
-	waiting_sender: mpsc::Sender<Response>,
-	waiting_response: mpsc::Receiver<Response>,
+	state: RpcHandlerState<M>,
+	is_options: bool,
+	cors_header: cors::CorsHeader<header::AccessControlAllowOrigin>,
+}
+
+impl<M: Metadata, S: Middleware<M>> Future for RpcHandler<M, S> {
+	type Item = server::Response;
+	type Error = hyper::Error;
+
+	fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+		let new_state = match mem::replace(&mut self.state, RpcHandlerState::Done) {
```

Comment:
We want to take the ownership of some stuff stored in the current state (like `ReadingHeaders { request, cors_domains, ...}`).